In [2]:
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sb
from sklearn.preprocessing import StandardScaler, PowerTransformer, QuantileTransformer, MinMaxScaler, MaxAbsScaler, RobustScaler, power_transform
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
from sklearn.ensemble import RandomForestRegressor
import xgboost as xgb
from sklearn.svm import SVR
from sklearn.neighbors import KNeighborsRegressor

/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [5]:
# os.chdir('drive/My Drive/Datasets/')
os.listdir()

['bottle.csv',
 'heart.csv',
 'iris.csv',
 'pima-indians-diabetes.csv',
 'sonar.csv',
 'housing.csv',
 'winequality-white.csv',
 'auto.csv',
 'sales.csv',
 'shampoo.csv',
 'birth.csv',
 'daily-min-temperatures.csv']

In [6]:
dataset = pd.read_csv('bottle.csv')

/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (47,73) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [7]:
dataset.head()

Cst_Cnt  Btl_Cnt       Sta_ID  ... pH2  pH1  DIC Quality Comment
0        1        1  054.0 056.0  ... NaN  NaN                  NaN
1        1        2  054.0 056.0  ... NaN  NaN                  NaN
2        1        3  054.0 056.0  ... NaN  NaN                  NaN
3        1        4  054.0 056.0  ... NaN  NaN                  NaN
4        1        5  054.0 056.0  ... NaN  NaN                  NaN

[5 rows x 74 columns]

In [8]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 864863 entries, 0 to 864862
Data columns (total 74 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   Cst_Cnt              864863 non-null  int64  
 1   Btl_Cnt              864863 non-null  int64  
 2   Sta_ID               864863 non-null  object 
 3   Depth_ID             864863 non-null  object 
 4   Depthm               864863 non-null  int64  
 5   T_degC               853900 non-null  float64
 6   Salnty               817509 non-null  float64
 7   O2ml_L               696201 non-null  float64
 8   STheta               812174 non-null  float64
 9   O2Sat                661274 non-null  float64
 10  Oxy_µmol/Kg          661268 non-null  float64
 11  BtlNum               118667 non-null  float64
 12  RecInd               864863 non-null  int64  
 13  T_prec               853900 non-null  float64
 14  T_qual               23127 non-null   float64
 15  S_prec           

In [9]:
nulls = dataset.isnull().sum().sort_values(ascending = False)
prcet = round(nulls/len(dataset)*100,2)

dataset_null = pd.DataFrame(columns =  ['Attr','Total','Percent'])
dataset_null.Attr  = nulls.index
dataset_null.Total = nulls.values
dataset_null.Percent = prcet.values
print(dataset_null.head(20))

                   Attr   Total  Percent
0                   pH2  864853   100.00
1   DIC Quality Comment  864808    99.99
2                   pH1  864779    99.99
3                  DIC2  864639    99.97
4                   TA2  864629    99.97
5                  DIC1  862864    99.77
6                   TA1  862779    99.76
7                C14A2p  852121    98.53
8                C14A1p  852103    98.52
9                C14As2  850449    98.33
10               C14As1  850431    98.33
11               IncTim  850426    98.33
12               LightP  846212    97.84
13               DarkAp  844406    97.63
14               MeanAp  844406    97.63
15               DarkAs  842214    97.38
16               MeanAs  842213    97.38
17               T_qual  841736    97.33
18                NH3uM  799901    92.49
19                R_NH4  799881    92.49


In [10]:
for i in dataset_null.Attr[dataset_null['Percent'] > 30]:
    dataset = dataset.drop([i], axis=1)

In [11]:
dataset.isnull().sum()

Cst_Cnt             0
Btl_Cnt             0
Sta_ID              0
Depth_ID            0
Depthm              0
T_degC          10963
Salnty          47354
O2ml_L         168662
STheta          52689
O2Sat          203589
Oxy_µmol/Kg    203595
RecInd              0
T_prec          10963
S_prec          47354
P_qual         191108
Chlqua         225697
Phaqua         225693
NH3q            56564
C14A1q          16258
C14A2q          16240
DarkAq          24423
MeanAq          24424
R_Depth             0
R_TEMP          10963
R_POTEMP        46047
R_SALINITY      47354
R_SIGMA         52856
R_SVA           52771
R_DYNHT         46657
R_O2           168662
R_O2Sat        198415
R_PRES              0
dtype: int64

In [12]:
for i in dataset.columns:
    if dataset[i].isnull().sum() > 0:
        dataset[i].fillna(dataset[i].mode().mean(), inplace=True)

In [13]:
dataset.isnull().sum()

Cst_Cnt        0
Btl_Cnt        0
Sta_ID         0
Depth_ID       0
Depthm         0
T_degC         0
Salnty         0
O2ml_L         0
STheta         0
O2Sat          0
Oxy_µmol/Kg    0
RecInd         0
T_prec         0
S_prec         0
P_qual         0
Chlqua         0
Phaqua         0
NH3q           0
C14A1q         0
C14A2q         0
DarkAq         0
MeanAq         0
R_Depth        0
R_TEMP         0
R_POTEMP       0
R_SALINITY     0
R_SIGMA        0
R_SVA          0
R_DYNHT        0
R_O2           0
R_O2Sat        0
R_PRES         0
dtype: int64

In [14]:
dataset.head(20)

Cst_Cnt  Btl_Cnt       Sta_ID  ... R_O2  R_O2Sat  R_PRES
0         1        1  054.0 056.0  ...  5.8    102.4       0
1         1        2  054.0 056.0  ...  5.8    102.4       8
2         1        3  054.0 056.0  ...  5.8    102.4      10
3         1        4  054.0 056.0  ...  5.8    102.4      19
4         1        5  054.0 056.0  ...  5.8    102.4      20
5         1        6  054.0 056.0  ...  5.8    102.4      30
6         1        7  054.0 056.0  ...  5.8    102.4      39
7         1        8  054.0 056.0  ...  5.8    102.4      50
8         1        9  054.0 056.0  ...  5.8    102.4      58
9         1       10  054.0 056.0  ...  5.8    102.4      75
10        1       11  054.0 056.0  ...  5.8    102.4      78
11        1       12  054.0 056.0  ...  5.8    102.4     101
12        1       13  054.0 056.0  ...  5.8    102.4     118
13        1       14  054.0 056.0  ...  5.8    102.4     126
14        1       15  054.0 056.0  ...  5.8    102.4     151
15        1       16  054.0 056.0  ...  5.8    102.4     153
16        1       17  054.0 056.0  ...  5.8    102.4     201
17        1       18  054.0 056.0  ...  5.8    102.4     222
18        1       19  054.0 056.0  ...  5.8    102.4     252
19        1       20  054.0 056.0  ...  5.8    102.4     287

[20 rows x 32 columns]

In [15]:
dataset['Depth_ID'].value_counts()

20-16045S-MX-110-2018-06330900-0025A-3    2
20-16045S-MX-110-1421-06670900-0026A-3    2
20-16045S-MX-110-2018-06330900-0002A-3    2
20-16045S-MX-110-2018-06330900-0010A-3    2
20-16045S-MX-110-2018-06330900-0088A-3    2
                                         ..
19-6901AX-CT-027-1707-11670350-0030A-5    1
19-5811PT-HY-321-1230-09330500-0400A-7    1
19-9704NM-MX-107-2342-07670900-0318A-3    1
19-6310BD-HY-282-2324-09330800-0250A-7    1
19-6504AX-HY-092-0312-09330600-0020A-7    1
Name: Depth_ID, Length: 864850, dtype: int64

In [16]:
dataset.dtypes

Cst_Cnt          int64
Btl_Cnt          int64
Sta_ID          object
Depth_ID        object
Depthm           int64
T_degC         float64
Salnty         float64
O2ml_L         float64
STheta         float64
O2Sat          float64
Oxy_µmol/Kg    float64
RecInd           int64
T_prec         float64
S_prec         float64
P_qual         float64
Chlqua         float64
Phaqua         float64
NH3q           float64
C14A1q         float64
C14A2q         float64
DarkAq         float64
MeanAq         float64
R_Depth        float64
R_TEMP         float64
R_POTEMP       float64
R_SALINITY     float64
R_SIGMA        float64
R_SVA          float64
R_DYNHT        float64
R_O2           float64
R_O2Sat        float64
R_PRES           int64
dtype: object

In [17]:
dataset['Sta_ID'].value_counts()

090.0 045.0    10043
090.0 070.0    10039
090.0 037.0     9771
090.0 060.0     9521
080.0 060.0     9393
               ...  
094.0 035.0        1
095.0 060.0        1
086.0 060.0        1
083.0 060.0        1
109.0 040.0        1
Name: Sta_ID, Length: 2634, dtype: int64

In [18]:
dataset.drop(['Depth_ID', 'Sta_ID'], axis=1, inplace=True)

In [19]:
dataset.dtypes

Cst_Cnt          int64
Btl_Cnt          int64
Depthm           int64
T_degC         float64
Salnty         float64
O2ml_L         float64
STheta         float64
O2Sat          float64
Oxy_µmol/Kg    float64
RecInd           int64
T_prec         float64
S_prec         float64
P_qual         float64
Chlqua         float64
Phaqua         float64
NH3q           float64
C14A1q         float64
C14A2q         float64
DarkAq         float64
MeanAq         float64
R_Depth        float64
R_TEMP         float64
R_POTEMP       float64
R_SALINITY     float64
R_SIGMA        float64
R_SVA          float64
R_DYNHT        float64
R_O2           float64
R_O2Sat        float64
R_PRES           int64
dtype: object

In [20]:
feature = dataset.drop(['T_degC'], axis=1)
label = dataset['T_degC']

In [21]:
def preproces(type_of_transform):
  scaler = type_of_transform

  feature_scaled = scaler.fit_transform(feature.values) 
  feature_train, feature_test, label_train, label_test = train_test_split(feature_scaled, label, test_size=0.25, random_state=10)
  
  lin_reg = LinearRegression()
  lin_reg.fit(feature_train, label_train)

  predictions = lin_reg.predict(feature_test)

  print("\nprediksi menggunakan ", type_of_transform)
  # print('Akurasi dari Testing Set: {:.3f}'.format(lin_reg.score(feature_test, label_test))) # .score() -> Accuracy
  print("r2 score")
  print(r2_score(label_test, predictions))
  print("RMSE")
  print(np.sqrt(mean_squared_error(label_test, predictions)))
  

In [61]:
preproces(StandardScaler())


prediksi menggunakan  StandardScaler(copy=True, with_mean=True, with_std=True)
r2 score
0.9999999166345492
RMSE
0.001220985085569301


In [62]:
preproces(PowerTransformer())

/usr/local/lib/python3.6/dist-packages/numpy/core/_methods.py:195: RuntimeWarning: overflow encountered in multiply
  x = um.multiply(x, x, out=x)
/usr/local/lib/python3.6/dist-packages/numpy/core/_methods.py:199: RuntimeWarning: overflow encountered in reduce
  ret = umr_sum(x, axis, dtype, out, keepdims)



prediksi menggunakan  PowerTransformer(copy=True, method='yeo-johnson', standardize=True)
r2 score
0.990149915052328
RMSE
0.41969838272829113


In [63]:
preproces(QuantileTransformer())


prediksi menggunakan  QuantileTransformer(copy=True, ignore_implicit_zeros=False, n_quantiles=1000,
                    output_distribution='uniform', random_state=None,
                    subsample=100000)
r2 score
0.9684540516138381
RMSE
0.7510850228864725


In [64]:
preproces(MinMaxScaler())


prediksi menggunakan  MinMaxScaler(copy=True, feature_range=(0, 1))
r2 score
0.9999999166345492
RMSE
0.0012209850855688332


In [65]:
preproces(MaxAbsScaler())


prediksi menggunakan  MaxAbsScaler(copy=True)
r2 score
0.9999999166345492
RMSE
0.0012209850855698075


In [66]:
preproces(RobustScaler())


prediksi menggunakan  RobustScaler(copy=True, quantile_range=(25.0, 75.0), with_centering=True,
             with_scaling=True)
r2 score
0.9999999166345492
RMSE
0.0012209850855692736


In [67]:
preproces(power_transform(feature,method='box-cox'))

ValueError: ignored

Untuk hasil dari beberapa metode transformasi nilai akurasi nya hampir sama semua untuk metode **StandardScaler(), RobustScaler(), MaxAbsScaler(), MinMaxScaler()** yaitu **0.9999**, sedangkan untuk metode **yeo-johnson** nilai akurasi nya sedikit lebih rendah yaitu **0.9901** dan nilai akurasi yang paling rendah didapatkan dengan menggunakan metode **QuantileTransformer()** yaitu **0.9684**.
Untuk Metode **Box Cox Transformation** mengalami error karena metode ini hanya bisa digunakan pada data yang positif sedangkan angka 0 tidak dapat benar benar dikatakan positif maupun negatif. .

In [23]:
scaler = StandardScaler()

feature_scaled = scaler.fit_transform(feature.values) 
feature_train, feature_test, label_train, label_test = train_test_split(feature_scaled, label, test_size=0.25, random_state=10)

In [71]:
#Random Forest Regressor
random_forest = RandomForestRegressor(n_estimators=50, bootstrap=False, max_features = "sqrt",max_depth=None, min_samples_split=2, min_samples_leaf=1, min_weight_fraction_leaf=0.0, max_leaf_nodes=None, random_state=10)
random_forest.fit(feature_train, label_train)

predictions = random_forest.predict(feature_train)

print("prediksi menggunakan training set")
# print('Akurasi dari Testing Set: {:.3f}'.format(lin_reg.score(feature_test, label_test))) # .score() -> Accuracy
print("r2 score")
print(r2_score(label_train, predictions))
print("RMSE")
print(np.sqrt(mean_squared_error(label_train, predictions)))

predictions = random_forest.predict(feature_test)

print("\nprediksi menggunakan testing set")
# print('Akurasi dari Testing Set: {:.3f}'.format(lin_reg.score(feature_test, label_test))) # .score() -> Accuracy
print("r2 score")
print(r2_score(label_test, predictions))
print("RMSE")
print(np.sqrt(mean_squared_error(label_test, predictions)))

prediksi menggunakan training set
r2 score
0.9999999999914501
RMSE
1.233516078819347e-05

prediksi menggunakan testing set
r2 score
0.9999136920468149
RMSE
0.039286443624797655


In [24]:
#XGBoost
xgb_model = xgb.XGBRegressor(max_depth=3, learning_rate=0.02, n_estimators=400, verbosity=1, objective='reg:squarederror', booster='gbtree',)
xgb_model.fit(feature_train, label_train)

predictions = xgb_model.predict(feature_train)

print("prediksi menggunakan training set")
# print('Akurasi dari Testing Set: {:.3f}'.format(lin_reg.score(feature_test, label_test))) # .score() -> Accuracy
print("r2 score")
print(r2_score(label_train, predictions))
print("RMSE")
print(np.sqrt(mean_squared_error(label_train, predictions)))

predictions = xgb_model.predict(feature_test)

print("\nprediksi menggunakan testing set")
# print('Akurasi dari Testing Set: {:.3f}'.format(lin_reg.score(feature_test, label_test))) # .score() -> Accuracy
print("r2 score")
print(r2_score(label_test, predictions))
print("RMSE")
print(np.sqrt(mean_squared_error(label_test, predictions)))

prediksi menggunakan training set
r2 score
0.9999954485977464
RMSE
0.008999877294293673

prediksi menggunakan testing set
r2 score
0.9999953970633569
RMSE
0.00907266355991354


In [26]:
#SVM Regressor
svm_reg = SVR(kernel='linear', C=10.0)
svm_reg.fit(feature_train, label_train)

predictions = svm_reg.predict(feature_train)

print("prediksi menggunakan training set")
# print('Akurasi dari Testing Set: {:.3f}'.format(lin_reg.score(feature_test, label_test))) # .score() -> Accuracy
print("r2 score")
print(r2_score(label_train, predictions))
print("RMSE")
print(np.sqrt(mean_squared_error(label_train, predictions)))

predictions = svm_reg.predict(feature_test)

print("\nprediksi menggunakan testing set")
# print('Akurasi dari Testing Set: {:.3f}'.format(lin_reg.score(feature_test, label_test))) # .score() -> Accuracy
print("r2 score")
print(r2_score(label_test, predictions))
print("RMSE")
print(np.sqrt(mean_squared_error(label_test, predictions)))

prediksi menggunakan training set
r2 score
0.999822931896306
RMSE
0.056135054193939084

prediksi menggunakan testing set
r2 score
0.9998239747762798
RMSE
0.05610541256220035


In [25]:
#KNN Regressor
knn = KNeighborsRegressor(n_neighbors=5, leaf_size=10, p=2, n_jobs=None, algorithm='brute')
knn.fit(feature_train, label_train)

predictions = knn.predict(feature_train)

print("prediksi menggunakan training set")
# print('Akurasi dari Testing Set: {:.3f}'.format(lin_reg.score(feature_test, label_test))) # .score() -> Accuracy
print("r2 score")
print(r2_score(label_train, predictions))
print("RMSE")
print(np.sqrt(mean_squared_error(label_train, predictions)))

predictions = knn.predict(feature_test)

print("\nprediksi menggunakan testing set")
# print('Akurasi dari Testing Set: {:.3f}'.format(lin_reg.score(feature_test, label_test))) # .score() -> Accuracy
print("r2 score")
print(r2_score(label_test, predictions))
print("RMSE")
print(np.sqrt(mean_squared_error(label_test, predictions)))

prediksi menggunakan training set
r2 score
0.9993643613694656
RMSE
0.10635774474443979

prediksi menggunakan testing set
r2 score
0.9990031705303154
RMSE
0.13351428738801677


Untuk model nya baru selesai **Random Forest** dan **XGBoost** dengan nilai **r2_score : 0.9999.** Untuk model algoritma yang lainnya membutuhkan waktu eksekusi yang lama, sehingga belum dapat dilihat hasil nya. 